compiler/frontend/src/print.rs

```diff
@@ -0,0 +1,304 @@
+}
+impl MaxLength {
+    fn fits(self, len: usize) -> bool {
+        match self {
+            MaxLength::Unlimited => true,
+            MaxLength::Limited(max) => len <= max,
+        }
+    }
+}
+
+/// Prints the value, using the visitor to match across possible values.
+///
+pub fn print<T>(
+    value: T,
+    precedence: Precedence,
+    max_length: MaxLength,
+    visitor: &impl Fn(T) -> Option<PrintValue<T>>,
+) -> Option<String> {
+    // For each case, the different alternatives of printing are listed.
+    // Depending on the available space, the best is chosen.
+    Some(match visitor(value)? {
+        PrintValue::Int(int) => {
+            // - int
+            // - `…`
+
+            let string = int.to_string();
+            if max_length.fits(string.len()) {
+                string
+            } else {
+                "…".to_string()
+            }
+        }
+        PrintValue::Tag { symbol, value } => {
+            // - full: `Tag Value` or `(Tag Value)` or `Tag`
+            // - only symbol: `Tag …` or `(Tag …)` or `Tag`
+            // - only structure: `… …` or `(… …)` or `…`
+            // - `…`
+
+            let needs_parentheses = value.is_some() && precedence == Precedence::High;
+
+            let length_needed_for_structure = match (needs_parentheses, value.is_some()) {
+                (false, false) => 1, // `…`
+                (false, true) => 3,  // `… …`
+                (true, false) => unreachable!(),
+                (true, true) => 5, // `(… …)`
+            };
+            if !max_length.fits(length_needed_for_structure) {
+                return Some("…".to_string());
+            }
+
+            let mut string = "".to_string();
+            if needs_parentheses {
+                string.push('(');
+            }
+
+            let symbol_fits = max_length.fits(length_needed_for_structure - 1 + symbol.len());
+            if symbol_fits {
+                string.push_str(&symbol);
+            } else {
+                string.push('…');
+            }
+
+            if let Some(value) = value {
+                string.push(' ');
+                if symbol_fits {
+                    let value = print(value, Precedence::High, max_length, visitor)?;
+                    if max_length.fits(length_needed_for_structure - 2 + symbol.len() + value.len())
+                    {
+                        string.push_str(&value);
+                    } else {
+                        string.push('…');
+                    }
+                } else {
+                    string.push('…');
+                }
+            }
+            if needs_parentheses {
+                string.push(')');
+            }
+            string
+        }
+        PrintValue::Text(text) => {
+            // - full text
+            // - `…`
+
+            let text = format!("\"{text}\"");
+            if max_length.fits(text.len()) {
+                text
+            } else {
+                "…".to_string()
+            }
```

Comment:
Avoid the formatting if we know it won't fit:

```suggestion
            if max_length.fits(1 + text.len() + 1) {
                format!("\"{text}\"")
            } else {
                "…".to_string()
            }
```